### DISCUSSION TITLE
Add explicit optional template variable

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#178] [Ideas]
When template errors are enabled for missing variables, it's difficult to work with optional values. This can be made easier by adding a feature to specify that a variable is optional, so it shouldn't throw error when missing.
Some ideas for syntax:
```jinja
{{ optional:possibly_missing | default:'text' }}

{# inspired by JS nullish coalescing operator #}
{{ possibly_missing ?? 'text' }}

{# second argument is default value #}
{% optional(possibly_missing, 'text') %}
```
I would prefer the first or second as they enable using other filters on the resulting value.

**Comment by ellmetha** [comment]
In strict variables mode, the best way to handle optional variables is really to ensure that those "optional" variables are initialized to `nil` in the template context from the handler or a context producer. 

Having the ability to define optional variables in templates is something that could make its way to the framework, but the proposed syntaxes appear inconsistent with the Marten templating language. Something that I think could eventually work would be the ability to define an optional variable (with an optional default value eventually) from a template using a new `optional` template tag.

```jinja
{# Assigns possibly_missing to nil if it is not defined #}
{% optional possibly_missing %}

{# Assigns possibly_missing to "text" if it is not defined #}
{% optional possibly_missing = "text" %}
```

Extending the capabilities of the `assign` template tag to allow something like that could also work:

```jinja
{# Assigns possibly_missing to nil if it is not defined #}
{% assign possibly_missing unless defined %}

{# Assigns possibly_missing to "text" if it is not defined #}
{% assign possibly_missing = "text" unless defined %}
```

In both cases, the important point is that it'd be mandatory to _explicitly_ define the optional variable _before_ using it in filter expressions.

**Reply by notramo** [reply]
These are better synaxes indeed. The `assign` based one would be definitely useful.
My use case is to have an input partial.
```jinja
<input type="{{ type | default:'text' }}">
```
I could add a context producer which sets these variables to `nil`, however, it can quickly become a source of bugs as it's not obvious where it was initialized. (That's why I suggested adding component support to the framework.)
Anyways, since this problem also affects other parts, not only components, I suggest adding a default context producer to the framework, for this purpose. Benefits would be:
- There was a convention for doing it, so the developers would know where to find it without reading any code. (As opposed to DIY solutions which differ from project to project.)
- The error message can suggest adding the missing variable in that context producer.

These could ensure that developers can learn the solution without figuring out by themselves (or opening support tickets like this), so they can focus on building the app.
